sources/cfgenv/source.go

```diff
@@ -0,0 +1,140 @@
+// Package cfgenv is a parameter provider that reads values from environment
+// variables.
+//
+// In order to be able to perform the strongest possible validation of the
+// provided environment variables, it is expected that all variables used
+// to configure an application have common prefix, and all variables with
+// that prefix must match a configuration parameter. This allows detecting
+// small mistakes like typos when specifying the name of variable.
+//
+// The rule for determining the environment variable that configures a
+// parameter is as follows:
+//
+//     {prefix}__{paramname}            (if parameter is not on a set)
+//     {prefix}__{setname}__{paramname} (if parameter is on a set)
+//     replace "-" by "_"
+//     uppercase
+//
+// Handling of booleans: the value of the environment variable must be a
+// valid go boolean, except for empty string, which is assumed to be "true".
+//
+// For example, is the prefix is "cfg":
+//
+//     param=test1-parameter           => env=CFG__TEST1_PARAMETER
+//     param=test2_parameter           => env=CFG__TEST2_PARAMETER
+//     param=address          set=http => env=CFG__HTTP__ADDRESS
+//
+// Note that both "-" and "_" are mapped to "_". For this reason, if one
+// application has two parameters that are differentiated only this letter,
```

Comment:
```suggestion
// application has two parameters that are differentiated only by this character,
```

Reply:
documentation aside, shouldn't this use case not be supported at all by the library and produce a panic immediately?